### PR TITLE
Leave archive join channel

### PR DESCRIPTION
**Pull request by ZaradarDFDS**

### CODE REVIEW

File: src/Harald.WebApi/Startup.cs [hunk 1]
```diff
@@ -23,7 +23,6 @@
 using Microsoft.Extensions.Diagnostics.HealthChecks;
 using Microsoft.Extensions.Hosting;
 using Prometheus;
-using IHostingEnvironment = Microsoft.AspNetCore.Hosting.IHostingEnvironment;
```
Comment (SEQUOIIA) [inline src/Harald.WebApi/Startup.cs]:
```
Infrastructure/Facades/Slack/SlackFacade.cs(10,32): error CS0246: The type or namespace name 'ISlackFacade' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(21,27): error CS0246: The type or namespace name 'CreateChannelResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(46,27): error CS0246: The type or namespace name 'SendNotificationResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(54,27): error CS0246: The type or namespace name 'SendNotificationResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(62,27): error CS0246: The type or namespace name 'SendNotificationResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(71,27): error CS0246: The type or namespace name 'GeneralResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(96,27): error CS0246: The type or namespace name 'CreateUserGroupResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(132,27): error CS0246: The type or namespace name 'GetConversationsResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(155,32): error CS0246: The type or namespace name 'UserGroup' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]
Infrastructure/Facades/Slack/SlackFacade.cs(174,80): error CS0246: The type or namespace name 'GeneralResponse' could not be found (are you missing a using directive or an assembly reference?) [harald/src/Harald.WebApi/Harald.WebApi.csproj]

The build failed. Please fix the build errors and run again.
```

File: src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        public static implicit operator string(SlackChannelName input)
+        {
+            return input.ToString();
+        }
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+        {
+            var results = new List<ValidationResult>();
+
+            if (_value.Length > 21)
```
Comment (SEQUOIIA) [inline src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs]: [(#30) 80 char slack channel name](https://github.com/dfds/harald/pull/30)

File: src/Harald.Tests/features/Connections/Infrastructure/DrivingAdapters/Api/TestConnectionsRoute.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                    .Build();
+
+                var response = await client.DeleteAsync($"api/v1/connections?channelId={channelId}&clientType={clientType}");
+
+                Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+            }
+        }
+
+        [Fact]
+        public async Task add_connection_returns_expected_status_code()
```
Comment (SEQUOIIA) [inline src/Harald.Tests/features/Connections/Infrastructure/DrivingAdapters/Api/TestConnectionsRoute.cs]:
```
Harald.Tests.Features.Connections.Infrastructure.DrivingAdapters.Api.TestConnectionsRoute.add_connection_returns_expected_status_code

System.NullReferenceException : Object reference not set to an instance of an object.
   at Harald.WebApi.Features.Connections.Infrastructure.DrivingAdapters.Api.ConnectionsController.AddConnection(ConnectionDto connection) in harald/src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs:line 117
   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at NSwag.AspNetCore.Middlewares.SwaggerUiIndexMiddleware`1.Invoke(HttpContext context)
   at NSwag.AspNetCore.Middlewares.RedirectToIndexMiddleware.Invoke(HttpContext context)
   at NSwag.AspNetCore.Middlewares.SwaggerDocumentMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.TestHost.HttpContextBuilder.<>c__DisplayClass10_0.<<SendAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.TestHost.ClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Harald.Tests.Features.Connections.Infrastructure.DrivingAdapters.Api.TestConnectionsRoute.add_connection_returns_expected_status_code() in harald/src/Harald.Tests/features/Connections/Infrastructure/DrivingAdapters/Api/TestConnectionsRoute.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
```

File: src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs (hunk 2 above)
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs]: I have changed this to support 80 chars or less. :)

File: src/Harald.Tests/features/Connections/Infrastructure/DrivingAdapters/Api/TestConnectionsRoute.cs (hunk 3 above)
Comment (ZaradarDFDS) [reply src/Harald.Tests/features/Connections/Infrastructure/DrivingAdapters/Api/TestConnectionsRoute.cs]: Done :)

File: src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs [hunk 4]
```diff
@@ -70,11 +82,102 @@ public class ConnectionsController : ControllerBase
...
+            
             var connectionDtos = connections.Select(ConnectionDto.CreateFromConnection);
 
             return Ok(new ItemsEnvelope<ConnectionDto>(connectionDtos));
         }
+
+        [HttpPost]
+        public async Task<IActionResult> AddConnection(ConnectionDto connection)
+        {
+            if (string.IsNullOrEmpty(connection.ClientId))
```
Comment (kim-lindhard-dfds) [inline src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs]: We could do these validations on the object factories and return 422 instead of the 400

File: src/Harald.WebApi/Features/Connections/Infrastructure/Persistence/FindConnectionsByClientTypeClientIdChannelTypeChannelIdHandler.cs [hunk 5]
```diff
@@ -23,30 +23,27 @@ public class FindConnectionsByClientTypeClientIdChannelTypeChannelIdHandler : IQ
...
-            if (query.ClientType != null && query.ClientType.GetType() != typeof(ClientTypeCapability))
+            if (query.ClientType != null && !query.ClientType.IsEmpty && !(query.ClientType is ClientTypeCapability))
             {
                 return Enumerable.Empty<Connection>();
             }
 
             var capabilities = await _capabilityRepository.GetAll();
 
-            if (query.ChannelId != null)
+            if (query.ChannelId != null && !query.ChannelId.IsEmpty)
```
Comment (kim-lindhard-dfds) [inline src/Harald.WebApi/Features/Connections/Infrastructure/Persistence/FindConnectionsByClientTypeClientIdChannelTypeChannelIdHandler.cs]: We could drop the IsEmpty check if we used the ChannelId object and made sure you couldn't give it a empty/space string

File: src/Harald.Infrastructure.Slack/ValidationMessages.resx [hunk 6]
```diff
@@ -0,0 +1,126 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/ValidationMessages.resx]: Yes :)

File: src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            return new SlackChannelName(input);
+        }
+
+        public static implicit operator string(SlackChannelName input)
+        {
+            return input.ToString();
+        }
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs]: At the end of the day slack will throw a validation error in our face if we try to create a channel with longer names. WIth that said I agree the validation should be performed in the facade before calling the method. So I will add in the appropriate logic to do so in the facade.

File: src/Harald.Infrastructure.Slack/Model/SlackChannelIdentifier.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            return new SlackChannelIdentifier(input);
+        }
+
+        public static implicit operator string(SlackChannelIdentifier input)
+        {
+            return input.ToString();
+        }
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/Model/SlackChannelIdentifier.cs]: At the time I was not sure if the ID was the name of the channel. But we can safely remove the validation logic here. :)

File: src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+            if (_value.Contains("_"))
+            {
+                results.Add(new ValidationResult(string.Format(ValidationMessages.Harald_Infrastructure_Slack_Model_SlackChannelName_Validate_Validate_Name, _value)));
+            }
+
+            return new List<ValidationResult>();
+        }
+
+        public override bool Equals(object obj)
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/Model/SlackChannelName.cs]: I simply copied the logic into the class from StringSubstitutable because I want to keep the domain out of the Slack facade assembly so its possible for us to package it up as a NuGet package and share it with others.

File: src/Harald.Infrastructure.Slack/Model/SlackChannelIdentifier.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            return new List<ValidationResult>();
+        }
+
+        public override bool Equals(object obj)
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.Infrastructure.Slack/Model/SlackChannelIdentifier.cs]: No :)

File: src/Harald.WebApi/Controllers/ChannelsController.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            catch(ValidationException exp)
+            {
+                return UnprocessableEntity(new {Message = exp.MessageToUser});
+            }
+
+            switch (convertedChannelType)
+            {
+                case ChannelTypeSlack _:
+                    return Accepted(await _slackFacade.GetChannels(_configuration["SLACK_API_AUTH_TOKEN"]));
+
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Controllers/ChannelsController.cs]: I would suggest you read the documentation from the Slack API. We need to add in a token parameter in the call to conversations.list in order to make sure that we scope the returned channels correctly. So this is not about adding in an auth token, its simply the way Slack decided to filter channels / conversations.

[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Controllers/ChannelsController.cs]: https://api.slack.com/methods/conversations.list

File: src/Harald.WebApi/Domain/StringSubstitutable.cs [hunk 12]
```diff
@@ -1,9 +1,13 @@
+using Microsoft.EntityFrameworkCore.Query.ExpressionTranslators.Internal;
+
 namespace Harald.WebApi.Domain
 {
     public abstract class StringSubstitutable
     {
         private readonly string _value;
 
+        public bool IsEmpty => string.IsNullOrEmpty(_value);
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Domain/StringSubstitutable.cs]: We need it to test if the inner value of the StringSubstitutable is null or empty. If we remove it there is no longer a way to test this.

File: src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs [hunk 13]
```diff
@@ -70,11 +82,102 @@ public class ConnectionsController : ControllerBase
...
+            {
+                return BadRequest("ChannelName is required.");
+            }
+
+            var userGroup = await _slackService.EnsureUserGroupExists(connection.ClientName);
+            var capability = Capability.Create(Guid.Parse(connection.ClientId), connection.ClientName, connection.ChannelId, userGroup.Id);
+
+            await _capabilityRepository.Add(capability);
+
+            var response = await _slackFacade.JoinChannel(connection.ChannelName);
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs]: If the channel does not exist it will be auto created by the Slack API. See more here @ https://api.slack.com/methods/channels.join. I have added logic to box / toString the class for the added validation check. Keep in mind that we get this on the API side as well so we need to decide if we want to keep this logic in two places. :)

File: src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs (hunk 4 above)
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Features/Connections/Infrastructure/DrivingAdapters/Api/ConnectionsController.cs]: Personally I always keep anything related to the protocol (in this case HTTP) in the controller, as it has the responsibility of mapping input to/from the protocol and the status code is very much part of this in my world. Also if we start putting HttpStatusCode objects inside our core factory classes their dependencies will end up all over the place. So I would prefere leaving it as is and sticking to throwing exceptions or returning null from our factories.

File: src/Harald.WebApi/Features/Connections/Infrastructure/Persistence/FindConnectionsByClientTypeClientIdChannelTypeChannelIdHandler.cs (hunk 5 above)
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Features/Connections/Infrastructure/Persistence/FindConnectionsByClientTypeClientIdChannelTypeChannelIdHandler.cs]: After I fixed the explicit operator our null checks didnt fail when the inner value was null, so your right we could do without it. I decided to leave it in as long as its possible to factorize SubstitutableString objects with inner values that are null to avoid edge cases that would cause the logic to fail (like if someone instantiates a type without using the factory) and ensure that the code paths are not activated when the value we are filtering on is "nullorempty".

**Comment by ZaradarDFDS** [comment]
**SlackFacade needs a base url & token**
Comment: It was not in scope for me to fix this. There are pros and cons of injecting a HttpClient vs. having the class create its own. Pros == cleaner code base, cons == using a transient HttpClient vs. an instance pr. facade instance could potentially yield a performance hit. Regardless of what we decide it will require us to rewire our DI container (which we should do anyways).

**HttpClient factory**
Comment: That sounds like a good idea. It will enable us to have the pros of a cleaner code base and avoid the cons of spawning to many HttpClients. :)

**Channel vs conversation**
Comment: Until they elimintate the channel concept from their API I think we should keep them both in place. 

**SlackFacade IDistributedCache**
Comment: I dont agree this voilates the SRP. Its the job of the cache we inject to handle caching, its the job of the facade to figure out when it want to cache (and what), since this is not something the cache can do on its own and I dont really see how wrapping it in a decorator will help this. :)

**Consider command handlers**
Comment. Totally agree. But out of scope for this PR.

**BDD Integration tests**
Comment: Totally agree. But out of scope for this PR.

File: src/Harald.IntegrationTests/Facades/Slack/SlackFacadeTest.cs [hunk 14]
```diff
@@ -1,198 +1,157 @@
...
             // Assert
             Assert.True(createChannelResponse.Ok);
             Assert.Equal(channelName, createChannelResponse.Channel.Name);
             Assert.NotEmpty(createChannelResponse.Channel.Id.ToString());
-            
-            // For eventual cleanup
-            _data.UserChannelId = createChannelResponse.Channel.Id;
+
+            //Clean up integration test resources.
+            await sut.ArchiveChannel(createChannelResponse.Channel.Id);
```
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.IntegrationTests/Facades/Slack/SlackFacadeTest.cs]: If we start using the token filter when we actually fetch channels, then all the channels created by our integration test should never collide with anything and they will not bloat the output when we query channels/conversations in prod (granted that we use a different token in prod, then we do in test)

File: src/Harald.WebApi/Controllers/ChannelsController.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ZaradarDFDS) [reply src/Harald.WebApi/Controllers/ChannelsController.cs]: token is now optional. If the SlackFacade is provided with valid options the facade will use a "fallback value"
